device/sycl/src/clusterization/clusterization_algorithm.cpp

```diff
@@ -0,0 +1,126 @@
+    // buffer
+    vecmem::data::vector_buffer<std::size_t> cluster_prefix_sum(num_modules,
+                                                                m_mr.get());
+    vecmem::data::vector_buffer<std::size_t> clusters_per_module(num_modules,
+                                                                 m_mr.get());
+    copy.setup(cluster_prefix_sum);
+    copy.setup(clusters_per_module);
+
+    // Invoke the reduction kernel that gives the total number of clusters which
+    // will be found (and also computes the prefix sums and clusters per module)
+    auto total_clusters = vecmem::make_unique_alloc<unsigned int>(m_mr.get());
+    *total_clusters = 0;
+    traccc::sycl::clusters_sum(cells_per_event, sparse_ccl_indices,
+                               total_clusters, cluster_prefix_sum,
+                               clusters_per_module, m_mr.get(), m_queue);
+
+    // Vector of the exact cluster sizes, will be filled in cluster_counting
+    // kernel
+    vecmem::vector<unsigned int> cluster_sizes(*total_clusters, 0, &m_mr.get());
+    traccc::sycl::cluster_counting(
+        num_modules, sparse_ccl_indices, vecmem::get_data(cluster_sizes),
+        cluster_prefix_sum, cells_max, m_mr.get(), m_queue);
+
+    // copy the sizes to the std::vector to construct the buffer for component
+    // connection
+    std::vector<std::size_t> cluster_sizes_host(*total_clusters);
+    for (std::size_t i = 0; i < *total_clusters; ++i) {
+        cluster_sizes_host[i] = cluster_sizes[i];
+    }
+    // Cluster container buffer for the clusters and headers (cluster ids)
+    cluster_container_buffer clusters_buffer{
+        {*total_clusters, m_mr.get()},
+        {std::vector<std::size_t>(*total_clusters, 0), cluster_sizes_host,
+         m_mr.get()}};
+
+    copy.setup(clusters_buffer.headers);
+    copy.setup(clusters_buffer.items);
+
+    // Component connection kernel
+    traccc::sycl::component_connection(clusters_buffer, cells_per_event,
+                                       sparse_ccl_indices, cluster_prefix_sum,
+                                       cells_max, m_mr.get(), m_queue);
+
+    // Copy the sizes of clusters per each module to the std vector for
+    // measurement buffer initialization
+    std::vector<std::size_t> clusters_per_module_host;
+    copy(clusters_per_module, clusters_per_module_host);
+
+    // Resizable buffer for the measurements
+    measurement_container_buffer measurement_buffer{
+        {num_modules, m_mr.get()},
+        {std::vector<std::size_t>(num_modules, 0), clusters_per_module_host,
+         m_mr.get()}};
+    copy.setup(measurement_buffer.items);
+
+    // Measurement creation kernel
+    traccc::sycl::measurement_creation(measurement_buffer, clusters_buffer,
+                                       *total_clusters, m_queue);
+
+    // Copy the results back to the host
+    host_measurement_container measurements(&m_mr.get());
+    copy(measurement_buffer.items, measurements.get_items());
```

Comment:
I have a quick question; how long does this line take?

Reply:
That's around 0.012 s